src/sync_device.rs

```diff
@@ -0,0 +1,707 @@
+                        code: typ.0,
+                        value: value as _,
+                    });
+                    let state = SyncState::$var {
+                        time: $time,
+                        start: $typ(typ.0 + 1),
+                    };
+                    return Some((ev, state));
+                }
+            }
+        }
+    };
+}
+impl Device {
+    fn compensate_keys(&self, time: libc::timeval, start: Key) -> Option<(InputEvent, SyncState)> {
+        try_compensate!(
+            self,
+            time,
+            start,
+            Key,
+            KEY,
+            Keys,
+            supported_keys,
+            AttributeSet<Key>,
+            |st| st.key_vals().unwrap(),
+            |vals, key| vals.contains(key)
+        );
+        self.compensate_absolutes(time, AbsoluteAxisType(0))
+    }
+
+    fn compensate_absolutes(
+        &self,
+        time: libc::timeval,
+        start: AbsoluteAxisType,
+    ) -> Option<(InputEvent, SyncState)> {
+        try_compensate!(
+            self,
+            time,
+            start,
+            AbsoluteAxisType,
+            ABSOLUTE,
+            Absolutes,
+            supported_absolute_axes,
+            &[libc::input_absinfo],
+            |st| st.abs_vals().unwrap(),
+            |vals, abs| vals[abs.0 as usize].value
+        );
+        self.compensate_switches(time, SwitchType(0))
+    }
+
+    fn compensate_switches(
+        &self,
+        time: libc::timeval,
+        start: SwitchType,
+    ) -> Option<(InputEvent, SyncState)> {
+        try_compensate!(
+            self,
+            time,
+            start,
+            SwitchType,
+            SWITCH,
+            Switches,
+            supported_switches,
+            AttributeSet<SwitchType>,
+            |st| st.switch_vals().unwrap(),
+            |vals, sw| vals.contains(sw)
+        );
+        self.compensate_leds(time, LedType(0))
+    }
+
+    fn compensate_leds(
+        &self,
+        time: libc::timeval,
+        start: LedType,
+    ) -> Option<(InputEvent, SyncState)> {
+        try_compensate!(
+            self,
+            time,
+            start,
+            LedType,
+            LED,
+            Leds,
+            supported_leds,
+            AttributeSet<LedType>,
+            |st| st.led_vals().unwrap(),
+            |vals, led| vals.contains(led)
+        );
+        None
+    }
+}
+
+struct FetchEventsSynced<'a> {
+    dev: &'a mut Device,
+    range: std::ops::Range<usize>,
+    consumed_to: usize,
+    sync: SyncState,
+}
+
+enum SyncState {
+    None,
+    Keys {
+        time: libc::timeval,
+        start: Key,
+    },
+    Absolutes {
+        time: libc::timeval,
+        start: AbsoluteAxisType,
+    },
+    Switches {
+        time: libc::timeval,
+        start: SwitchType,
+    },
+    Leds {
+        time: libc::timeval,
+        start: LedType,
+    },
+}
+
+impl<'a> Iterator for FetchEventsSynced<'a> {
```

Comment:
This is complicated enough I feel it needs some unit tests with happy path cases for cases with a block containing SYN_DROPPED and the expected output, along with a validation that we don't return "partial" blocks until we see a SYN_REPORT at the end (because we may need to drop them). Again, this is complex because it's possible for event_buf to contain incomplete blocks at the end, or at the beginning if we don't do this correctly.